server/server.go

```diff
@@ -0,0 +1,297 @@
+}
+
+var _ PrincipalParser = (*principalParser)(nil)
+
+type context struct {
+	id                    principal.Signer
+	canIssue              CanIssueFunc
+	principal             PrincipalParser
+	validateAuthorization AuthorizationValidatorFunc
+}
+
+func (ctx *context) ID() principal.Signer {
+	return ctx.id
+}
+
+func (ctx *context) CanIssue(capability ucan.Capability[any], issuer did.DID) bool {
+	return ctx.canIssue(capability, issuer)
+}
+
+func (ctx *context) Principal() PrincipalParser {
+	return ctx.principal
+}
+
+func (ctx *context) ValidateAuthorization(auth validator.Authorization) result.Failure {
+	return ctx.validateAuthorization(auth)
+}
+
+var _ InvocationContext = (*context)(nil)
+
+type server struct {
+	id      principal.Signer
+	service Service
+	context InvocationContext
+	codec   transport.InboundCodec
+	catch   ErrorHandlerFunc
+}
+
+func (srv *server) ID() principal.Signer {
+	return srv.id
+}
+
+func (srv *server) Service() Service {
+	return srv.service
+}
+
+func (srv *server) Context() InvocationContext {
+	return srv.context
+}
+
+func (srv *server) Codec() transport.InboundCodec {
+	return srv.codec
+}
+
+func (srv *server) Request(request transport.HTTPRequest) (transport.HTTPResponse, error) {
+	return Handle(srv, request)
+}
+
+func (srv *server) Run(invocation ServiceInvocation) (receipt.AnyReceipt, error) {
+	return Run(srv, invocation)
+}
+
+func (srv *server) Catch(err HandlerExecutionError[any]) {
+	srv.catch(err)
+}
+
+var _ transport.Channel = (*server)(nil)
+var _ ServerView = (*server)(nil)
+
+func Handle(server Server, request transport.HTTPRequest) (transport.HTTPResponse, error) {
+	selection, aerr := server.Codec().Accept(request)
+	if aerr != nil {
+		return thttp.NewHTTPResponse(aerr.Status(), strings.NewReader(aerr.Error()), aerr.Headers()), nil
+	}
+
+	msg, err := selection.Decoder().Decode(request)
+	if err != nil {
+		return thttp.NewHTTPResponse(http.StatusBadRequest, strings.NewReader("The server failed to decode the request payload. Please format the payload according to the specified media type."), nil), nil
+	}
+
+	result, err := Execute(server, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return selection.Encoder().Encode(result)
+}
+
+func Execute(server Server, msg message.AgentMessage) (message.AgentMessage, error) {
+	br, err := blockstore.NewBlockReader(blockstore.WithBlocksIterator(msg.Blocks()))
+	if err != nil {
+		return nil, err
+	}
+
+	var invs []invocation.Invocation
+	for _, invlnk := range msg.Invocations() {
+		inv, err := invocation.NewInvocationView(invlnk, br)
+		if err != nil {
+			return nil, err
+		}
+		invs = append(invs, inv)
+	}
+
+	var rcpts []receipt.AnyReceipt
+	var rerr error
+	var wg sync.WaitGroup
+	var lock sync.RWMutex
+	for _, inv := range invs {
+		wg.Add(1)
+		go func(inv invocation.Invocation) {
+			defer wg.Done()
+			rcpt, err := Run(server, inv)
+			if err != nil {
+				rerr = err
+				return
+			}
+
+			lock.Lock()
+			defer lock.Unlock()
```

Comment:
general note: if it's just one line with no conditionals , there's no need for a defer -- just put it right after.

NB